chore: List widget functionalities; DB connections and binding

chandannkumar opened the pull request:
## Description

Added functionalities to bind list widget with Data sources like Postgres and API

## Type of change


## How Has This Been Tested?

- Cypress


## Checklist:
### Dev activity
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] PR is being merged under a feature flag


### QA activity:
- [ ] Test plan has been approved by relevant developers
- [ ] Test plan has been peer reviewed by QA
- [ ] Cypress test cases have been added and approved by either SDET or manual QA
- [ ] Organized project review call with relevant stakeholders after Round 1/2 of QA
- [ ] Added Test Plan Approved label after reveiwing all Cypress test

chandannkumar commented:
/ok-to-test sha=460d14e

Aishwarya-U-R commented on app/client/cypress/integration/Smoke_TestSuite/ClientSideTests/Widgets/List/ListWidgetFunctionality_Spec.ts:
```diff
@@ -0,0 +1,244 @@
+    ee.SelectEntityByName("Image1")
+   // agHelper.GetNClick(locator._imageWidget, 0, true) //Image1
+    propPane.UpdatePropertyFieldValue("Image", "{{currentItem.imageUrl}}")
+    deployMode.DeployApp()
```
Deploying the App & doing what? Add validations here.

chandannkumar replied:
Added new validations

Aishwarya-U-R commented on app/client/cypress/integration/Smoke_TestSuite/ClientSideTests/Widgets/List/ListWidgetFunctionality_Spec.ts:
```diff
@@ -0,0 +1,244 @@
+        ee.SelectEntityByName("Image1")
+        //agHelper.GetNClick(locator._imageWidget, 0, true) //Image1
+        propPane.UpdatePropertyFieldValue("Image", "{{currentItem.avatar}}")
+        deployMode.DeployApp()
```
Simply moving to Deploy mode. Add validations to all following cases too.

chandannkumar replied:
Made some modification in Spec and added validations in all cases

Aishwarya-U-R commented on app/client/cypress/integration/Smoke_TestSuite/ClientSideTests/Widgets/List/ListWidgetFunctionality_Spec.ts:
```diff
@@ -0,0 +1,244 @@
+    
+    })
+})
+
```
In the afterAll() - We usually delete the created datasources/queries/apis. But here is not needed. This should be added wisely.

chandannkumar replied:
Noted, Thanks

chandannkumar commented:
/ok-to-test sha=823d4f3

chandannkumar commented on app/client/cypress/integration/Smoke_TestSuite/ClientSideTests/Widgets/List/ListWidgetFunctionality_Spec.ts:
```diff
@@ -0,0 +1,244 @@
+    ERROR_ACTION_EXECUTE_FAIL,
+    createMessage,
+  } from "../../../../../support/Objects/CommonErrorMessages";
+import { EntityExplorer } from "../../../../../support/Pages/EntityExplorer";
```
Sure

chandannkumar commented:
/ok-to-test sha=79b4e32

chandannkumar commented:
/ok-to-test=0630910

chandannkumar commented:
/ok-to-test sha=0630910

chandannkumar commented:
/ok-to-test sha=70096f4

chandannkumar commented:
/ok-to-test sha=f5918bf

chandannkumar commented:
Thanks @Aishwarya-U-R & @Vijetha-Kaja for reviewing this. I have made all changes mentioned in callouts.
Also, I have added validations on Spec file, added few common locators and added a method in Aggregate Helper

chandannkumar commented:
/ok-to-test sha=559677a

chandannkumar commented on app/client/cypress/integration/Smoke_TestSuite/ClientSideTests/Widgets/List/ListWidgetFunctionality_Spec.ts:
```diff
@@ -0,0 +1,244 @@
+import { EntityExplorer } from "../../../../../support/Pages/EntityExplorer";
+import { first, propertyOf } from "lodash";
+import { data } from "cypress/types/jquery";
+import { listenerCount, on } from "process";
```
Done

chandannkumar commented:
/ok-to-test sha=703a0a3

Vijetha-Kaja commented on app/client/cypress/integration/Smoke_TestSuite/ClientSideTests/Widgets/List/ListWidgetFunctionality_Spec.ts:
```diff
@@ -0,0 +1,178 @@
+after(() => {
+    _.ee.ExpandCollapseEntity("Queries/JS")
+    _.ee.ActionContextMenuByEntityName("API1","Delete","Are you sure?")
+    _.ee.ActionContextMenuByEntityName("postgres_ssp","Delete","Are you sure?")
```
Please add deletion of the created datasource also

Aishwarya-U-R replied:
We have other specs validating this DS deletion behavior. So here it is not mandatory. But just remember it as a standard.

chandannkumar replied:
Noted, Since here i was unable to delete the DS due to intend behavior, so i skipped deleting it. 
Thanks for the information

Aishwarya-U-R commented on app/client/cypress/support/Pages/AggregateHelper.ts:
```diff
@@ -984,6 +984,16 @@ export class AggregateHelper {
     return cy.contains(text).should(exists);
   }
 
+  public AssertElementDisabled(
```
Not able to use - AssertElementEnabledDisabled() ? Is there a need for this new method? Can you please confirm?

chandannkumar replied:
AssertElementEnabledDisabled() has a parameter called index, for which List widget pagination was unable to find the exact position. To make it easier, created a new method to only assert the presence of element and should be disabled.

Aishwarya-U-R replied:
Made changes accordingly. Please refer to this file going forward for any of your new scripting.

Aishwarya-U-R commented on app/client/cypress/integration/Smoke_TestSuite/ClientSideTests/Widgets/List/ListWidgetFunctionality_Spec.ts:
```diff
@@ -0,0 +1,178 @@
+describe("Verify List widget binding & functionalities with Queries and API", function() {
+
+it("1. Create new API & verify data on List widget", function(){
+    const apiUrl = `https://thronesapi.com/api/v2/Characters`
```
Using external api. Should be more careful here.

chandannkumar replied:
Changed it to mock api of appsmith

Aishwarya-U-R commented on app/client/cypress/integration/Smoke_TestSuite/ClientSideTests/Widgets/List/ListWidgetFunctionality_Spec.ts:
```diff
@@ -0,0 +1,178 @@
+})
+
+
+it("7. Verify Pagination in SSP and verify no pagination visible when SSP is disabled", function(){
```
Whats SSP? - Expand in the title of the test case itelf.

chandannkumar replied:
Yes, sure

Aishwarya-U-R commented on app/client/cypress/integration/Smoke_TestSuite/ClientSideTests/Widgets/List/ListWidgetFunctionality_Spec.ts:
```diff
@@ -0,0 +1,178 @@
+        _.agHelper.GetNAssertElementText(_.locators._listPaginateActivePage,"2","have.text")
+        _.deployMode.NavigateBacktoEditor()
+        _.agHelper.GetNClick(_.locators._listWidget,1,true)
+        _.propPane.ToggleOnOrOff("Server Side Pagination", "Off")
+        _.agHelper.AssertElementAbsence(_.locators._listPaginateItem)
```
No Deploy page validation for Server Side Pagination - Off?

chandannkumar replied:
Added same validation to verify on deploy mode as well

chandannkumar commented:
/ok-to-test sha=1e7e920

chandannkumar commented:
/ok-to-test sha=b07b173

Aishwarya-U-R commented on app/client/cypress.json:
```diff
@@ -1,5 +1,5 @@
 {
-	"baseUrl": "https://dev.appsmith.com/",
+	"baseUrl": "https://release.app.appsmith.com/",
```
Why is cypress.json merged? Remove it.

Aishwarya-U-R commented on app/client/cypress/fixtures/datasources.json:
```diff
@@ -5,17 +5,33 @@
   "mongo-databaseName": "mongo_samples",
```
Properly pull the latest on this file & run. No local changes should be merged for this file too, remove it ASAP.

chandannkumar commented:
/ok-to-test sha=18bb75f

Aishwarya-U-R commented on app/client/cypress/integration/Regression_TestSuite/ClientSideTests/Widgets/List/ListBasic_Spec.ts:
```diff
@@ -0,0 +1,221 @@
+
+describe("Verify List widget binding, Server side Pagination & functionalities with Queries and API", function() {
+  it("1. Create new API & verify data on List widget", function() {
+    const apiUrl = `https://mock-api.appsmith.com/users`;
```
We are decommissioning the live api soon. Use mock-api from TED.
Take help from @Vijetha-Kaja if needed for the same.

chandannkumar replied:
Updated

chandannkumar commented:
/ok-to-test sha=94939b3

chandannkumar commented:
/ok-to-test sha=cdf4ece

Aishwarya-U-R commented:
/ok-to-test sha=a82fb61

chandannkumar commented:
/ok-to-test sha=9614e53

chandannkumar commented:
/ok-to-test sha=be56599

chandannkumar commented:
/ok-to-test sha=c436246

chandannkumar commented:
/ok-to-test sha=737df7a

chandannkumar commented:
/ok-to-test sha=503d872

chandannkumar commented:
/ok-to-test sha=f1e375b
